### DISCUSSION TITLE
reference data disappearing after pytest run?

### DISCUSSION THREAD

**Discussion by ccurvey** [pytest-dev/pytest#12992] [Q&A]
I'm sure this is something I've done; I'm hoping the community can help me figure out where I went wrong.

I'm using pytest with Django.  In my migrations, I create a Category class, and populate it with reference data.   I have `--reuse-db` in my `pytest.ini`.   

A few days ago, some of my tests started failing because some of my Category reference data was missing.  I can get my tests to pass if I specify `--create-db` on the command line.  However, after a test run, the Category table exists in Postgres, but it's empty.

Anyone seen something like this before?  Any ideas on how I might be able to track down the issue?

**Comment by portyu9** [comment]
Hi @ccurvey 

The key detail here is:

```text
--reuse-db
```

This is pytest-django behavior rather than pytest core.

`--reuse-db` deliberately keeps the test database between runs for faster startup. It does **not** mean that database contents should be treated as persistent application/reference state between test runs.

If you changed migrations or migration-created data, the existing reused test database can be stale. In that situation use:

```bash
pytest --reuse-db --create-db
```

`--create-db` forces pytest-django to rebuild the test database, after which subsequent runs can reuse it again.

That explains why:

```bash
pytest --create-db
```

makes the tests pass.

There is a second issue to consider, though: tests using Django's database support perform database isolation/cleanup. Data that a test modifies or creates should not be expected to remain in the reused database after the suite.

So I would not rely on:

```text
migration inserts Category rows once
        ↓
reused test DB permanently keeps them
        ↓
all future pytest runs consume them
```

as your test-data strategy.

If those categories are required for a test, establish that precondition deterministically. Depending on what the data represents, that can be:

- a data migration when the database is created,
- a pytest fixture that creates the required objects,
- or pytest-django's database setup customization for global test-database initialization.

For example, for test-specific reference data:

```python
@pytest.fixture
def categories(db):
    return [
        Category.objects.create(name="A"),
        Category.objects.create(name="B"),
    ]
```

Then tests that require those categories request the fixture explicitly.

For your immediate problem I would do two things:

1. After any migration/reference-data change, rebuild once:

```bash
pytest --reuse-db --create-db
```

2. Inspect the tests for anything deleting/flushing `Category` rows, because an empty table after the suite means some test/database cleanup path has removed the records.

So `--reuse-db` is useful for keeping the expensive database/schema around, but it should not be treated as a mechanism for persisting mutable test data indefinitely.

Please mark this answer as accepted if it helped, thank you.
